horaemeta/pkg/coderr/error.go

```diff
@@ -25,82 +25,135 @@ import (
 	error
 	Code() Code
-	// WithCausef should generate a new CodeError instance with the provided cause details.
-	WithCausef(format string, a ...any) CodeError
+}
+
+// CodeErrorDef is an error definition, and able to generate a real CodeError
+type CodeErrorDef interface {
+	Code() Code
+	// WithMessagef should generate a new CodeError instance with the provided message.
+	WithMessagef(format string, a ...any) CodeError
 	// WithCause should generate a new CodeError instance with the provided cause details.
 	WithCause(cause error) CodeError
+	// WithCausef should generate a new CodeError instance with the provided cause details and error message.
+	WithCausef(cause error, format string, a ...any) CodeError
 }
 
 // Is checks whether the cause of `err` is the kind of error specified by the `expectCode`.
 // Returns false if the cause of `err` is not CodeError.
 func Is(err error, expectCode Code) bool {
-	code, b := GetCauseCode(err)
-	if b && code == expectCode {
+	code, ok := GetCauseCode(err)
+	if ok && code == expectCode {
 		return true
 	}
 
 	return false
 }
 
-func GetCauseCode(err error) (Code, bool) {
-	if err == nil {
-		return Invalid, false
-	}
+// codeError is the default implementation of CodeError.
+type codeError struct {
+	code           Code
+	msg            string
+	causeWithStack error
+	hasCause       bool
+}
 
-	cause := errors.Cause(err)
-	cerr, ok := cause.(CodeError)
-	if !ok {
-		return Invalid, false
+func (e *codeError) Error() string {
+	if e.hasCause {
+		return fmt.Sprintf("[err_code=%d]%s, cause:%v", e.code, e.msg, e.causeWithStack)
+	} else {
+		return fmt.Sprintf("[err_code=%d]%s", e.code, e.msg)
 	}
-	return cerr.Code(), true
 }
 
-// NewCodeError creates a base CodeError definition.
-// The provided code should be defined in the code.go in this package.
-func NewCodeError(code Code, desc string) CodeError {
-	return &codeError{
-		code:  code,
-		desc:  desc,
-		cause: nil,
-	}
+func (e *codeError) Code() Code {
+	return e.code
 }
 
-// codeError is the default implementation of CodeError.
-type codeError struct {
-	code  Code
-	desc  string
-	cause error
+// NewCodeErrorDef creates an CodeError definition.
+// The provided code should be defined in the code.go in this package.
+func NewCodeErrorDef(code Code, desc string) CodeErrorDef {
+	return &codeErrorDef{
+		code: code,
+		desc: desc,
+	}
 }
 
-func (e *codeError) Error() string {
-	return fmt.Sprintf("(#%d)%s, cause:%+v", e.code, e.desc, e.cause)
+// codeErrorDef is the default implementation of CodeErrorDef.
+type codeErrorDef struct {
+	code Code
+	desc string
 }
 
-func (e *codeError) Code() Code {
+func (e *codeErrorDef) Code() Code {
 	return e.code
 }
 
-func (e *codeError) WithCausef(format string, a ...any) CodeError {
+func (e *codeErrorDef) WithMessagef(format string, a ...any) CodeError {
 	errMsg := fmt.Sprintf(format, a...)
-	causeWithStack := errors.WithStack(errors.New(errMsg))
+	causeWithStack := errors.WithStack(errors.New(""))
```

Comment:
IMO, this function should not add stack, users can use `WithCausef` to do this. Otherwise this function is same with `WithCausef`